Jellyfin audiobook download

lysanntranvouez opened the pull request:
## Purpose

Adds a feature "Download from Jellyfin", similar to "Download from URL".

- UI to log in to a jellyfin server
- Lists all audiobooks found on the server (all "book" libraries, and all folders & audiobooks in those)
- Selecting an audiobook starts the download process, hooking into the existing single file download

## Related tasks

- #1205

## Approach

- For the most part it doesn't touch existing features, only adds new views, view models & controllers
- Only exception: Extracted single file download (by URL) into a separate service, so it is reused for the jellyfin download

## Things to be aware of / Things to focus on

- I haven't programmed Swift or iOS for 8(?) years. I have never done SwiftUI before.
- I tried to stay similar to the existing patterns I spotted (presentation flows, use of MVVM), but I'm not super familiar with how things are done on iOS these days, so there's likely room for improvement.

## Draft status

This is still very much a draft at this stage:
- [x] UX has varying amounts of quality
- [x] There is zero error handling
- [x] There are no unit tests

Other missing things before I consider this ready for merge:
* Must have:
  - [x] Option to remember login information + add option to sign out (forget login info) in settings (and probably the jellyfin download view itself?)
  - [x] Document the added 3rd party libs
* Should have:
  - [x] A bit of a question - I think it would be good to have some kind of a page _before_ the download, to see the size + a download button. So that the user initiates the download very intentionally. Or else, books and folders need to be distinguished better.

The below would be nice to have, but I think we can skip them for the initial PR. It's quite chonky already.

* Nice to have:
  - [ ] Ability to queue several downloads
  - [ ] Search function in the jellyfin library

Super optional but nice to have:
- [ ] Ability to download an entire folder and save the files into a folder with the same name

## Screenshots

Simulator recording of an earlier version: 
https://github.com/user-attachments/assets/39b15e1e-301f-4073-a2bd-0e7e22486b21

GianniCarlo reviewed (commented):
Overall this looks very promising, I've been wrapping my head around the SDK and how it works, and it got me excited as a potential first step into integrating other services as well, replacing the 'Add' import sheet, for a screen with the options, and registered / connected services.

I left a few comments, and I think that if you split off the refactoring of the single download service, we can merge that other PR in quite easily.

I haven't looked into the majority of the SwiftUI code yet in the Jellyfin folder, as I was more focusing on the coordinator and how the SDK was authorizing the user, but I wanted to leave this feedback as it's already been a week since the PR was created. I think the grid UI works, and am not very picky on the interface, as its an 'external' list of books

Re streaming, I think this should be fairly easy, I see that from the jellyfin item URL that is used to download the book, we only need to add the `api_key` with the access token as a parameter. I think if we pass the book info back to the app, and create the `Book` with that url assigned to the property `remoteURL`. We would have to remove the check `syncService.isActive` in the `PlayerManager.loadPlayerItem` function (well or replace it with something better to account for books on an external service), and it should automagically works. This could be addressed in a second PR though, to avoid adding more into this one

lysanntranvouez commented on BookPlayer/Jellyfin/JellyfinCoordinator.swift:
```diff
@@ -0,0 +1,82 @@
+  private let singleFileDownloadService: SingleFileDownloadService
+  private var disposeBag = Set<AnyCancellable>()
+
+  private var apiClient: JellyfinClient?
+  private var userID: String?
+  private var libraryName: String?
```
I'll have another look at this.

On that note, I planned to keep it simple and have just one saved connection for simplicity. If other services are ever integrated, it could make more sense to allow several jellyfin connections too, since it would just be one more connection in a list that needs to be managed anyway. But I think for this PR I would rather keep it simple. But as I said, I'll have another look at those optionals.

GianniCarlo replied:
keeping the scope to just one connection for this pr sounds good, just wondering about the logout to that connection, is that meant to be handled in a separate PR?

lysanntranvouez replied:
I plan to add them to this PR.

lysanntranvouez commented on BookPlayer/en.lproj/Localizable.strings:
```diff
@@ -319,3 +320,15 @@ We're working hard on providing a seamless experience, if possible, please conta
 "Rewind ${interval}" = "Rewind ${interval}";
 "settings_lock_orientation_title" = "Orientation Locked";
 "more_title" = "More";
+"jellyfin_connection_title" = "Jellyfin";
```
I was wondering how it works, do I add just the Base strings? Base + English?

GianniCarlo replied:
yeah that should be enough, just so I feed that into Lokalise.com and get the translation keys from there and update the other files

lysanntranvouez commented on BookPlayer/AppDelegate.swift:
```diff
@@ -135,13 +135,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate, BPLogger {
         playbackService: playbackService,
         playerManager: playerManager
       )
+      let singleFileDownloadService = SingleFileDownloadService(
```
Okay, that makes sense. I was very unsure about what should go in there, or rather, where to initialize stuff that isn't a "core" service. Thank you for the feedback.

lysanntranvouez replied:
It's updated now. Let me know if this is what you imagined (and note that I also added a new JellyfinConnectionService, which gets passed in the same fashion).

lysanntranvouez commented:
Sorry for the late reply, I've been quite busy the last couple weeks.

> I left a few comments, and I think that if you split off the refactoring of the single download service, we can merge that other PR in quite easily.

Double checking if I understand you correctly. Do you want me to remove the single file download service refactoring from this PR, so that we can merge this PR? Or are you asking for something else?

I'm asking because without the refactoring I'd have to go back to the more basic version of the file download for jellyfin, which also means almost no user feedback. I'd rather not reimplement the user feedback you've already made for the download from URL flow.<br/>
But I can fix the issues with my refactoring based on your feedback, like how the service is registered.<br/>
Were there any other issues I need to address?

Let me know if I misunderstood anything or if you disagree.

lysanntranvouez commented:
Yeah I have some work in progress changes to store the credentials in the keychain.<br/>
I kind of shoehorned the data into the available fields... But I like your suggestion of a Codable structure better, so I'll adjust that.

Could you point me at your other branch? I had made another class to store my keychain data. But it would be much better to have one place to interface with the keychain, since its API is a bit... special :D

lysanntranvouez commented on BookPlayer/Library/ItemDetails Screen/ClearableTextField.swift:
```diff
@@ -14,18 +14,27 @@ struct ClearableTextField: View {
   /// Input's text
   @Binding var text: String
+
+  public var keyboardType: UIKeyboardType = .default
+  public var textContentType: UITextContentType? = nil
+  public var autocapitalization: UITextAutocapitalizationType = .sentences
```
Riiight, okay. I feel very dumb now.

One question though: Do I still need to pass in some value/flag to decide about `TextField` vs `SecureField`?

I had this:
```
  @ViewBuilder
   private var textField: some View {
     if self.textContentType == .password {
       SecureField(placeholder, text: $text)
     } else {
       TextField(placeholder, text: $text)
     }
   }
```

How can I achieve that without adding a property to `ClearableTextField`?

Or would you rather I make a `ClearableSecureField`? Or that I just use a `SecureField`?

GianniCarlo replied:
hmmm good point, we could have a ClearableSecureField, though I'm not sure if it's really needed, meaning that a SecureField once it loses and regain focus, the next key press from the user clears it already. Hmmm on second thought, let's just use SecureField directly, as it keeps in line with the general UI from the OS

GianniCarlo commented:
Re my branch, I haven't pushed it yet, but I'll split off my changes into a separate branch and push that, will ping back when that's up (at the latest on Monday, not sure if I'll be able to get to it tomorrow)

GianniCarlo commented:
@lysanntranvouez I put up my changes for the keychain service [here](https://github.com/TortugaPower/BookPlayer/pull/1219/files#diff-f4b745f5c9e5059a78afdbc37544d8ada7260c0f051c4f0a813974f14023d84aR21), I'll merge it once the tests run, and the idea would be to just add a new key to the enum, and use the get / set functions with the codable struct, let me know if you have any questions (Re automockable, Sourcery was giving problems because of the overload with the generic so I just removed it)

lysanntranvouez commented:
@GianniCarlo 

Thank you!

> let me know if you have any questions (Re automockable, Sourcery was giving problems because of the overload with the generic so I just removed it)

When the unit tests run, do they use an isolated environment, or can they pick up values stored in the keychain on the target device/simulator when we don't use a mock?

GianniCarlo commented:
@lysanntranvouez yeah I was worried a bit about that, as the AccountService tests also make use of the keychain service, I honestly don't have any assurances that the keychain is not shared between runs even when parallelization is disabled for the tests (I did try that at some point, but it didn't play nice with the Github actions), but as soon as one of the tests fails, I'll probably end up creating a manual mock for the service

lysanntranvouez commented:
I've updated the PR with the ability to save the connection, and to remove the saved connection.

And I've addressed all of your comments, I think.

(But there's still more I want to do before considering this done.)

lysanntranvouez commented:
So I would say this is now a MVP. The important features are there, there are nice to haves, but I don't think it's the best idea to include them into this PR.

Updated demo video (simulator recording): https://github.com/user-attachments/assets/39b15e1e-301f-4073-a2bd-0e7e22486b21

-------

I am aware of one bug, and I need some help with that: Sometimes when I hit download, the app appears to do some kind of a tactical restart. The main UI seems to reload. I am not sure what's happening, or how to find out what's happening. Is this some kind of error handling in the app? Or am I somehow dismissing the root view controller or something?

Attaching another video to demonstrate what I mean: https://github.com/user-attachments/assets/923a3c49-1e33-4626-ac4c-aa9260479c7e

GianniCarlo commented:
Hi @lysanntranvouez ! sorry I've been a bit out this week, I'm almost done with the watch version (streaming is working, only need to add the download option), and I got tunnel vision.

Regarding the issue, you mentioned, the problem is here:

<img width="784" alt="Screenshot 2024-11-29 at 12 15 15 PM" src="https://github.com/user-attachments/assets/fcbd3e80-92eb-44d1-b724-3cd5521509f9">

There are instances where the `self.flow.finishPresentation(animated: true)` is being called more than once, so the first time the Jellyfin screens are dismissed, the second time, the dismiss function dismisses the main navigation controller since it's not presenting anything (this is just how UIKit handles it under the hood). You just have to make sure that a coordinator only finishes once

I can put it up on TestFlight so it's easier for anyone who wants to test this feature, let me know if you would rather I wait until you put in the fix, or if I should go ahead and put it up on TestFlight

Re reviewing the PR, I want to finish up the work on the Apple watch (I can almost feel it 😅) and I'll circle back to this PR

lysanntranvouez commented:
Don't worry, I wasn't done anyway, and it's a lot. I understand that this takes time.

Plus, that help was quick help, thank you so much! I just pushed a fix for that.

TestFlight would be great! (Can I have an invite to that?)

GianniCarlo commented:
Sure @lysanntranvouez ! I just put it up here: https://testflight.apple.com/join/ccKCfquZ (the link is open to anyone that wants to join an test it), I created a separate test group on TestFlight, so right now that build has only what is included in your PR

GianniCarlo commented:
@lysanntranvouez I'm finally done with my PR, I went over this one, and besides translations, I think it's in a pretty good shape just to merge in, and get feedback from Jellyfin users on the next release. Let me know how things are looking, as I mean to submit the new release by the end of this week

lysanntranvouez commented:
Hey there, sorry for the late reply - I was a bit under the weather.

After doing some proper on device testing I just pushed two tiny tweaks.

But, it's ready to go from my point of view :)

GianniCarlo commented:
@lysanntranvouez no worries! I went ahead and merged the PR, I did some tests and there was a small retain cycle with the `JellyfinConnectionViewModel` (I fixed [this in here](https://github.com/TortugaPower/BookPlayer/commit/00c5421a953ff36d099084794e05dcf345f2f821)), the thing is that under the current logic, the idea is that the Coordinator is kept in memory while the screen is in the navigation stack, and that's accomplished with the onTransition block from the ViewModel that keeps a reference to self from the coordinator, the problem was that on that transition block, the view model was being passed inside also, and the view model was keeping a reference to itself and leaking after the screen is dismissed.

I'll add the translations later today, and put up the final beta on all the three groups currently running betas, before submitting this weekend. I really appreciate your contribution with this PR 🙌, this really was a complex task to tackle
